docs: Add Custom memory limit resources (#83)
reanahub/docs.reana.io: docs/advanced-usage/compute-backends/kubernetes/index.md

        commands:
            - python "${helloworld}"
```

## Custom memory limit

> Kubernetes allows configuring and managing resources for containers. With REANA, you can configure cluster wide **memory** limits to make the effiecient use of available resources.

Reviewer: The setting to configure cluster-wide memory limits is just for admins and is already explained in the [Helm README](https://github.com/reanahub/reana/tree/maint-0.7/helm/reana) as `kubernetes_jobs_memory_limit`. This is not relevant for the users so I'd omit this part and explain better that a user can customise the memory limit of each workflow step in order to avoid jobs from running out of memory.

> To set the memory limit of a job you must specify `kubernetes_memory_limit`.

Reviewer: Please specify that this needs to be set in the step specification.

> This can be set directly in `bytes` and you can read more about the expected memory values on [kubernetes official documentation](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-memory)

Reviewer: >  This can be set directly in `bytes`

Although this is possible, I don't think it's the most common approach as it's not human-readable. The user can also specify human-readable units, so I'd shorten this part.

> Note that jobs exceeding the memory limits, you will get out of memory (OOM) error and job gets [`OOMKilled`](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#requests-and-limits).

Reviewer: The approach should be the opposite. As a user, by default, you don't touch anything, and if your job is so heavy that you get `OOMKilled`, then you have the possibility to increase the memory limit to avoid this error.

Reviewer: Then you can also mention that one can also decrease the memory limit to increase the chances of their jobs being scheduled earlier.

> ```yaml hl_lines="7"

Reviewer: Please add some text to introduce the examples, and mention which file it is in each case.

>    # Serial example

Reviewer: We can use normal markdown titles for this instead, and in each example you can explain a bit where the hint should be placed.

   ...
   steps:
      - name: reana_demo_helloworld_memory_limit
        environment: 'python:2.7-slim'
        compute_backend: kubernetes

>         kubernetes_memory_limit: '10Mi'

Reviewer: Maybe something more realistic like `8Gi`? I'm not sure what's our default memory limit but I think it's `6Gi`

        commands:
            - python helloworld.py
```

> ```yaml hl_lines="18"

Reviewer: Wrong line highlighted.

To introduce this example you can explain that the hin must be in the Yadage spec file under `step` > `environment` > `resources`...

   # Yadage example
   ...
   stages:
   - name: reana_demo_helloworld_memory_limit
      dependencies: [init]
      scheduler:
        scheduler_type: 'singlestep-stage'
        parameters:
          helloworld: {step: init, output: helloworld}
        step:
          process:
            process_type: 'string-interpolated-cmd'
            cmd: 'python "{helloworld}"'
          environment:
            environment_type: 'docker-encapsulated'
            image: 'python'
            imagetag: '2.7-slim'
            resources:
              - compute_backend: kubernetes
              - kubernetes_memory_limit: '10Mi'
```

> ```yaml hl_lines="10"

Reviewer: Wrong line highlighted.

Same as yadage.

        commands:
            - python "${helloworld}"
```

## Custom memory limit

> When jobs exceeds the memory limits, you will get out of memory (OOM) error and job gets [`OOMKilled`](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#requests-and-limits). With REANA, you can configure cluster wide **memory** limits to make the efficient use of available resources. Note that you can also decrease the memory limit to increase the chances of jobs being scheduled earlier.

Reviewer: ```suggestion
When jobs exceed the memory limits, you will get out of memory (OOM) error and job gets [`OOMKilled`](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#requests-and-limits). With REANA, you can configure cluster wide **memory** limits to make the efficient use of available resources. Note that you can also decrease the memory limit to increase the chances of jobs being scheduled earlier.
```

Reviewer: > With REANA, you can configure cluster wide **memory** limits to make the efficient use of available resources.

This is not relevant for the end-user, only for admins.

Here we should mention something like..

```
In order to avoid this error,  you can increase the **memory** limits per workflow step to make an efficient use of available resources. 
```

> When jobs exceeds the memory limits, you will get out of memory (OOM) error and job gets [`OOMKilled`](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#requests-and-limits). With REANA, you can configure cluster wide **memory** limits to make the efficient use of available resources. Note that you can also decrease the memory limit to increase the chances of jobs being scheduled earlier.

Reviewer: ```suggestion
When jobs exceeds the memory limits, you will get out of memory (OOM) error and job gets [`OOMKilled`](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#requests-and-limits). With REANA, you can configure cluster wide **memory** limits to make the efficient use of available resources. Note that you can also decrease the memory limit to increase the chances of jobs being scheduled earlier if the cluster is busy.
```

> When jobs exceed the memory limits, you will get out of memory (OOM) error and job gets [`OOMKilled`](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#requests-and-limits). To avoid OOM error, you can increase the **memory** limits per workflow step to make an efficient use of the available resources. Note that you can also decrease the memory limit to increase the chances of jobs being scheduled earlier if the cluster is busy.

Reviewer: ```suggestion
When jobs exceed the cluster memory limits, you will get out of memory (OOM) error and the job gets [`OOMKilled`](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#requests-and-limits). To avoid OOM error, you can increase the **memory** limits per workflow step to make an efficient use of the available resources. Note that you can also decrease the memory limit to increment the chances of jobs being scheduled earlier if the cluster is busy.
```

> To set the memory limit of a job you must specify `kubernetes_memory_limit` in _**every step specification**_.

Reviewer: I wouldn't say **every** as it implies that you're force to set it for all the steps.

```
To set the memory limit of a job you can specify `kubernetes_memory_limit` in _**each step specification**_.
```

> To set the memory limit of a job you must specify `kubernetes_memory_limit` in _**each step specification**_.

Reviewer: ```suggestion
To set the memory limit of a job you can specify `kubernetes_memory_limit` in the specification of **each  workflow step**.
```

> You can read more about the expected memory values on [kubernetes official documentation](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-memory)

Reviewer: ```suggestion
You can read more about the expected memory values on [Kubernetes official documentation](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-memory)
```

> You can read more about the expected memory values on [Kubernetes official documentation](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-memory)

Reviewer: ```suggestion
Read more about the expected memory values on [Kubernetes official documentation](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-memory).
```
To avoid "You can..." repetition

> You can configure the `steps` in respective yaml for serial, yadage and cwl workflow engines.

Reviewer: ```suggestion
You can configure the `steps` in respective yaml for Serial, Yadage and CWL workflow engines.
```

> You can configure the `steps` in respective yaml for Serial, Yadage and CWL workflow engines.

Reviewer: ```suggestion
You can configure the `steps` in respective YAML for Serial, Yadage and CWL workflow engines.
```

Author: I applied Diego's suggestion here. https://github.com/reanahub/docs.reana.io/pull/83#discussion_r621927552

> You can configure the `steps` in respective yaml for Serial, Yadage and CWL workflow engines.

Reviewer 2: ```suggestion
You can configure the `steps` in the respective `yaml` specifications for Serial, Yadage, and CWL workflow engines.
```

### Serial workflow

You can set `kubernetes_memory_limit` in every `step` of workflow specification.

```yaml hl_lines="6"
   ...
   steps:
      - name: reana_demo_helloworld_memory_limit
        environment: 'python:2.7-slim'
        compute_backend: kubernetes
        kubernetes_memory_limit: '8Gi'
        commands:
            - python helloworld.py
```

### Yadage workflow

> You can set `kubernetes_memory_limit` in every `step` under `environment` >> `resources`.

Reviewer 2: Perhaps we can use a more standard way of writing YAML paths:
```suggestion
You can set `kubernetes_memory_limit` in every `step` under `environment.resources`.
```

```yaml hl_lines="19"
   ...
   stages:
   - name: reana_demo_helloworld_memory_limit
      dependencies: [init]
      scheduler:
        scheduler_type: 'singlestep-stage'
        parameters:
          helloworld: {step: init, output: helloworld}
        step:
          process:
            process_type: 'string-interpolated-cmd'
            cmd: 'python "{helloworld}"'
          environment:
            environment_type: 'docker-encapsulated'
            image: 'python'
            imagetag: '2.7-slim'
            resources:
              - compute_backend: kubernetes
              - kubernetes_memory_limit: '8Gi'
```

### CWL workflow

> You can set `kubernetes_memory_limit` in every `step` under `hints` >> `reana`.

Reviewer 2: ```suggestion
You can set `kubernetes_memory_limit` in every `step` under `hints.reana`.
```
